service/src/main/java/com/kodemon/service/facade/FightFacadeImpl.java

```diff
@@ -0,0 +1,128 @@
+package com.kodemon.service.facade;
+
+import com.kodemon.api.dto.FightDTO;
+import com.kodemon.api.dto.GymDTO;
+import com.kodemon.api.dto.UserDTO;
+import com.kodemon.api.enums.WildPokemonFightMode;
+import com.kodemon.api.facade.FightFacade;
+import com.kodemon.persistence.entity.*;
+import com.kodemon.service.interfaces.*;
+import com.kodemon.service.util.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.inject.Inject;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * @author Oliver Roch
+ */
+
+@Service
+@Transactional
+public class FightFacadeImpl implements FightFacade {
+    @Inject
+    private BeanMappingService beanMappingService;
+
+    @Inject
+    private TrainerFightService trainerFightService;
+
+    @Inject
+    private PokemonFightService pokemonFightService;
+
+    @Inject
+    private PokemonService pokemonService;
+
+    @Inject
+    private TrainerService trainerService;
+
+    @Inject
+    private BadgeService badgeService;
+
+    @Inject
+    private TimeService timeService;
+
+
+    @Override
+    public void fightForBadge(UserDTO user, GymDTO gym) {
+        Trainer challengingTrainer = beanMappingService.mapTo(user, Trainer.class);
+        Gym targetGym = beanMappingService.mapTo(gym, Gym.class);
+
+        boolean wasChallengerSuccessful = false;
+        if(trainerFightService.wasFightForBadgeSuccessful(challengingTrainer, targetGym.getTrainer())) {
+            Badge badge = badgeService.createBadgeOfGym(targetGym);
+            trainerService.addBadge(badge, challengingTrainer);
+            badgeService.assignTrainerToBadge(challengingTrainer, badge);
```

Comment:
Swap those two calls, please, because you first need to save the badge to the database.